src/Sylius/Bundle/CartBundle/Controller/CartItemController.php

```diff
@@ -53,49 +39,61 @@ public function addAction(Request $request)
+        $form = $this->resourceFormFactory->create($configuration, $newResource);
 
-            return $this->redirectAfterAdd($configuration);
-        }
+        if ($request->isMethod('POST') && $form->submit($request)->isValid()) {
+            $newResource = $form->getData();
 
-        $event = new CartItemEvent($cart, $item);
+            $event = $this->eventDispatcher->dispatchPreEvent(ResourceActions::CREATE, $configuration, $newResource);
 
-        // Update models
-        $eventDispatcher->dispatch(SyliusCartEvents::ITEM_ADD_INITIALIZE, $event);
-        $eventDispatcher->dispatch(SyliusCartEvents::CART_CHANGE, new GenericEvent($cart));
-        $eventDispatcher->dispatch(SyliusCartEvents::CART_SAVE_INITIALIZE, $event);
+            if ($event->isStopped() && !$configuration->isHtmlRequest()) {
+                throw new HttpException($event->getErrorCode(), $event->getMessage());
+            }
+            if ($event->isStopped()) {
+                $this->flashHelper->addFlashFromEvent($configuration, $event);
 
-        // Write flash message
-        $eventDispatcher->dispatch(SyliusCartEvents::ITEM_ADD_COMPLETED, new FlashEvent());
+                return $this->redirectHandler->redirectToIndex($configuration, $newResource);
+            }
 
-        return $this->redirectAfterAdd($configuration);
-    }
+            $cart = $this->getCurrentCart();
+            $this->resolveCartItem($cart, $newResource);
 
-    /**
-     * Redirect to specific URL or to cart.
-     *
-     * @param Request $request
-     *
-     * @return RedirectResponse
-     */
-    private function redirectAfterAdd(RequestConfiguration $configuration)
-    {
-        $request = $configuration->getRequest();
+            $this->eventDispatcher->dispatchPostEvent(ResourceActions::CREATE, $configuration, $newResource);
+
+            if (!$configuration->isHtmlRequest()) {
+                return $this->viewHandler->handle($configuration, View::create($newResource, Response::HTTP_CREATED));
+            }
+
+            $translatedMessage = $this->translateMessage('sylius.cart.item_add_completed', $this->metadata);
+            $this->addFlash('success', $translatedMessage);
 
-        if ($request->query->has('_redirect_to')) {
-            return $this->redirectHandler->redirect($configuration, $request->query->get('_redirect_to'));
+            $this->recalculateCart($cart);
+
+            return $this->redirectHandler->redirectToResource($configuration, $newResource);
         }
 
-        return $this->redirectToCartSummary($configuration);
+        if (!$configuration->isHtmlRequest()) {
+            return $this->viewHandler->handle($configuration, View::create($form, Response::HTTP_BAD_REQUEST));
+        }
+
+        $view = View::create()
+            ->setData([
+                'configuration' => $configuration,
+                'metadata' => $this->metadata,
```

Comment:
This is a cart specific action, so I think we should not pass the metadata here.